File: code/design-patterns/creational/prototype/go/prototype.go
package prototype

import (
	"fmt"
	"math"
)

// Prototype interface
type Shape interface {
	Clone() Shape
	GetArea() float64
	GetSides() int
}

// Concrete prototype
type Circle struct {
	Radius int
}

func (c *Circle) Clone() Shape {
	// Create a new instance of Circle and copy all fields
	return &Circle{Radius: c.Radius}
}

func (c *Circle) GetArea() float64 {
	return math.Pi * float64(c.Radius) * float64(c.Radius)
}

func (c *Circle) GetSides() int {
	return 0
}

func (c *Circle) String() string {
	return fmt.Sprintf("Circle with radius %d", c.Radius)
}

type Rectangle struct {
	Width, Height int
}

func (r *Rectangle) Clone() Shape {
	// Create a new instance of Rectangle and copy all fields
	return &Rectangle{Width: r.Width, Height: r.Height}
}

func (r *Rectangle) GetArea() float64 {
	return float64(r.Width * r.Height)
}

func (r *Rectangle) GetSides() int {
	return 4
}

func (r *Rectangle) String() string {
	return fmt.Sprintf("Rectangle with width %d and height %d", r.Width, r.Height)
}

File: code/design-patterns/python/creational/prototype/prototype.py
from abc import ABC, abstractmethod
import math


# Prototype interface
class Shape(ABC):
    @abstractmethod
    def clone(self):
        pass

    @abstractmethod
    def get_area(self):
        pass

    @abstractmethod
    def get_sides(self):
        pass


# Concrete prototype
class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def clone(self):
        # Create a new instance of Circle and copy all fields
        new_circle = Circle(self.radius)
        return new_circle

    def get_area(self):
        return math.pi * self.radius**2

    def get_sides(self):
        return 0

    def __str__(self):
        return f"Circle with radius {self.radius}"


class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def clone(self):
        # Create a new instance of Rectangle and copy all fields
        new_rectangle = Rectangle(self.width, self.height)
        return new_rectangle

    def get_area(self):
        return self.width * self.height

    def get_sides(self):
        return 4

    def __str__(self):
        return f"Rectangle with width {self.width} and height {self.height}"
